src/effects/deeplearning/ModelCard.cpp

```diff
@@ -0,0 +1,272 @@
+******************************************************************/
+
+#include "ModelCard.h"
+#include "DeepModel.h"
+
+#include <rapidjson/document.h>
+#include <rapidjson/ostreamwrapper.h>
+#include <rapidjson/filewritestream.h>
+#include <rapidjson/writer.h>
+#include <rapidjson/filereadstream.h>
+#include <rapidjson/error/en.h>
+
+
+namespace validators 
+{
+
+   void validateExists(const std::string &key, DocHolder doc)
+   {
+      if(!doc->IsObject())
+         throw InvalidModelCardDocument("provided document is not an object", doc);
+      if(!doc->HasMember(key.c_str()))
+      {
+         std::stringstream msg;
+         msg<<"JSON document missing field: "<<key<<std::endl;
+         throw InvalidModelCardDocument(msg.str(), doc);
+      }
+   }
+
+   void throwTypeError(const std::string &key, const char *type, DocHolder doc)
+   {
+      std::stringstream msg;
+      msg << "field: " << key << " is not of type: " << type;
+      throw InvalidModelCardDocument(msg.str(), doc);
+   }
+
+   std::string tryGetString(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsString())
+         throwTypeError(key, "string", doc);
+
+      return (*doc)[key.c_str()].GetString();
+   }
+
+   int tryGetInt(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsInt())
+         throwTypeError(key, "int", doc);
+      
+      return (*doc)[key.c_str()].GetInt();
+   }
+
+   bool tryGetBool(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsBool())
+         throwTypeError(key, "bool", doc);
+      
+      return (*doc)[key.c_str()].GetBool();
+   }
+
+   std::vector<std::string> tryGetStringArray(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsArray())
+         throwTypeError(key, "array", doc);
+      
+      std::vector<std::string> labels;
+      for (auto itr = (*doc)[key.c_str()].Begin(); itr != (*doc)[key.c_str()].End();  ++itr)
+         labels.emplace_back(itr->GetString());
+
+      return labels;
+   }
+
+}
+
+namespace parsers 
+{
+   DocHolder ParseString(const std::string &data)
+   {
+      DocHolder d = std::make_shared<rapidjson::Document>();
+      // parse the data
+      d->Parse(data.c_str());
+      if (d->Parse(data.c_str()).HasParseError()) 
+      {
+         std::string msg = "error parsing JSON from string: \n";
+         msg += rapidjson::GetParseError_En(d->GetParseError());
+         msg += "\n document: \n\t" + std::string(data.c_str()) + "\n";
+         throw InvalidModelCardDocument(msg.c_str(), d);
+      }
+      
+      return d;
+   }
+
+   DocHolder ParseFile(const std::string &path)
+   {
+      // TODO: error check reading from the stream?
+      // TODO: error check for file parse
+      // https://rapidjson.org/md_doc_stream.html
+      FILE* fp = fopen(path.c_str(), "r");
```

Comment:
Please use wxFIle API, they will handle encoding correctly. This is a major pain point on Windows

Reply:
resolved [here](https://github.com/audacity/audacity/pull/1384/commits/ab4a59dbcf320a54d8b88e230abb2c4caf6e7bee)